### PR TITLE
Add support for dict/tuple obs space for VecCheckNaN

### LINKED ISSUE #1319: [Bug]: VecCheckNan with Dict observation space throws exception
### 🐛 Bug

Wrapping an environment with VecCheckNan that uses a Dict observation space results in an exception. Stack trace below.

There's no special logic for handling a Dict in https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/vec_env/vec_check_nan.py

### To Reproduce

```python
import gym
import numpy as np
from stable_baselines3.common.vec_env import DummyVecEnv, VecCheckNan


class MyEnv(gym.Env):
    def __init__(self):
        self.observation_space = gym.spaces.Dict(
            test=gym.spaces.Box(low=-np.inf, high=np.inf, shape=(8, 10)),
            test_discrete=gym.spaces.Discrete(2)
        )
        self.action_space = gym.spaces.Discrete(3)

    def reset(self):
        return {'test': np.zeros((8, 10)), 'test_discrete': 1}

    def step(self, action):
        return {'test': np.zeros((8, 10)), 'test_discrete': 1}, 0, False, {}


env = DummyVecEnv([lambda: MyEnv()])
env = VecCheckNan(env)

env.reset()

```

### Relevant log output / Error message

```shell
Traceback (most recent call last):
  File "C:\Users\Nate\rl-ge\rl-ge\repro.py", line 24, in <module>
    env.reset()
  File "C:\Users\Nate\rl-ge\env\lib\site-packages\stable_baselines3\common\vec_env\vec_check_nan.py", line 43, in reset
    self._check_val(event="reset", observations=observations)
  File "C:\Users\Nate\rl-ge\env\lib\site-packages\stable_baselines3\common\vec_env\vec_check_nan.py", line 54, in _check_val
    has_nan = np.any(np.isnan(val))
TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

### System Info

- OS: Windows-10-10.0.19044-SP0 10.0.19044
- Python: 3.9.16                          
- Stable-Baselines3: 1.7.0                
- PyTorch: 1.13.1+cpu                     
- GPU Enabled: False                      
- Numpy: 1.24.1                           
- Gym: 0.21.0

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] I have provided a minimal working example to reproduce the bug
- [X] I've used the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

### PR DESCRIPTION
**Pull request by araffin**
## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

closes #1319

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
Note: by activating the check in the env checker, the new code is run
- [ ] I have updated the documentation accordingly.
- [ ] I have opened an associated PR on the [SB3-Contrib repository](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib) (if necessary)
- [ ] I have opened an associated PR on the [RL-Zoo3 repository](https://github.com/DLR-RM/rl-baselines3-zoo) (if necessary)
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: docs/misc/changelog.rst [hunk 1]
```diff
@@ -18,6 +18,7 @@ New Features:
 ^^^^^^^^^^^^^
 - Added ``repeat_action_probability`` argument in ``AtariWrapper``.
 - Only use ``NoopResetEnv`` and ``MaxAndSkipEnv`` when needed in ``AtariWrapper``
+- Added suport for dict/tuple observations spaces for ``VecCheckNan``, the check is now active in the ``env_checker()`` (@DavyMorgan)
```
Comment (qgallouedec) [inline docs/misc/changelog.rst:21]: support

File: stable_baselines3/common/vec_env/vec_check_nan.py [hunk 2]
```diff
@@ -44,19 +48,40 @@ def reset(self) -> VecEnvObs:
...
+            if isinstance(value, (np.ndarray, list)):
+                found += self.check_array_value(name, np.asarray(value))
+            elif isinstance(value, dict):
+                for inner_name, inner_val in value.items():
+                    found += self.check_array_value(f"{name}.{inner_name}", inner_val)
+            elif isinstance(value, tuple):
+                for idx, inner_val in enumerate(value):
+                    found += self.check_array_value(f"{name}.{idx}", inner_val)
+            else:
+                raise ValueError(f"Unsupported observation type {type(value)}.")
```
Comment (qgallouedec) [inline stable_baselines3/common/vec_env/vec_check_nan.py:84]: I'd rather use TypeError

**Comment by qgallouedec** [comment]
The following now fails while checking info

```python
import gym
import numpy as np
from gym import spaces

from stable_baselines3.common.vec_env import DummyVecEnv, VecCheckNan


class MyEnv(gym.Env):
    def __init__(self):
        self.observation_space = spaces.Box(low=-np.inf, high=np.inf, shape=(8, 10))
        self.action_space = spaces.Discrete(3)

    def reset(self):
        return np.zeros((8, 10))

    def step(self, action):
        return np.zeros((8, 10)), 0, False, {}


env = DummyVecEnv([lambda: MyEnv()])
env = VecCheckNan(env)

env.reset()
env.step(env.step([env.action_space.sample()]))
```

```log
Traceback (most recent call last):
  File "/Users/quentingallouedec/stable-baselines3/1319.py", line 24, in <module>
    env.step(env.step([env.action_space.sample()]))
  File "/Users/quentingallouedec/stable-baselines3/stable_baselines3/common/vec_env/base_vec_env.py", line 162, in step
    self.step_async(actions)
  File "/Users/quentingallouedec/stable-baselines3/stable_baselines3/common/vec_env/vec_check_nan.py", line 35, in step_async
    self._check_val(event="step_async", actions=actions)
  File "/Users/quentingallouedec/stable-baselines3/stable_baselines3/common/vec_env/vec_check_nan.py", line 82, in _check_val
    found += self.check_array_value(f"{name}.{idx}", inner_val)
  File "/Users/quentingallouedec/stable-baselines3/stable_baselines3/common/vec_env/vec_check_nan.py", line 60, in check_array_value
    has_nan = np.any(np.isnan(value))
TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

**Comment by DavyMorgan** [comment]
It might be due to the mistake in the test script?
`env.step(env.step([env.action_space.sample()]))`

I think it has one extra step and should be
`env.step([env.action_space.sample()])`

**Comment by qgallouedec** [comment]
You're right @DavyMorgan.
LGTM then, just the minor comments

**Review by qgallouedec — APPROVED** [review]

**Pull request merged**